timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/client/DistributingModulusGeneratorTest.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.transaction.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+
+import com.google.common.collect.Iterables;
+
+public class DistributingModulusGeneratorTest {
+    private DistributingModulusGenerator generator;
+
+    @Test
+    public void doesNotReuseResiduesIfNotNeeded() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+    }
+
+    @Test
+    public void waitsToReuseResiduesIfNeeded() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+    }
+
+    @Test
+    public void throwsIfUnmarkingUnmarkedResidue() {
+        givenGeneratorHasModulus(3);
+
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void canUnmarkMarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResidues(3);
+        unmarkResidue(0);
+        unmarkResidue(1);
+        unmarkResidue(2);
+        // Not throwing is a success!
+    }
+
+    @Test
+    public void cannotDoubleUnmarkMarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResidues(3);
+
+        unmarkResidue(0);
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void reassignsUnmarkedResiduesIfTheyAreNowLeastReferenced() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        unmarkResidue(1);
+        assertThat(requestResidues(1)).containsExactly(1);
+    }
+
+    @Test
+    public void repeatedlyReassignsUnmarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+
+        unmarkResidue(2);
+        unmarkResidue(2);
+        unmarkResidue(2);
+        List<Integer> additionalResponses = requestResidues(6);
```

Comment:
do `requestResidues(3)` and then you can do the following check and `requestResiduesAndAssertBalanced(3);` without having to sublist